localstack/services/s3/provider_stream.py

```diff
@@ -131,11 +132,21 @@ def put_object(
         if content_sha_256 and content_sha_256.startswith("STREAMING-"):
             # this is a chunked request, we need to properly decode it while setting the key value
             decoded_content_length = int(headers.get("x-amz-decoded-content-length", 0))
-            key_object.set_value_from_chunked_payload(body, decoded_content_length)
+            # we're passing the S3 Object, the decoder will set its `ChecksumValue` once it has finished reading
+            # the stream
+            body = AwsChunkedDecoder(body, decoded_content_length, key_object)
 
-        else:
-            # set the stream to be the value of the key
+        try:
             key_object.value = body
```

Comment:
q: Hmm, why would setting the value here raise an exception?

Reply:
That's a good question 😄 basically, setting the value will trigger a property setter in moto, where it will verify the checksum. So it can trigger exception... in the new provider, I've decoupled the checksum logic, so there's a proper check after setting the value, but we're not there yet with this one.

Reply:
Okay, then maybe add a small comment explaining just that :)